File: packages/broadcast/test/conformance_test.dart
/// The Dart side of the shared wire vectors.
///
/// Two implementations parse this format — this client and the JavaScript
/// relay — and until these vectors existed each knew the layout only from
/// its own code. That is exactly how two implementations drift: nothing
/// fails until a deployment mixes an old reader with a new writer, in the
/// one situation where nobody can debug it.
///
/// The file is the format written down. If a change here fails, either the
/// wire format changed on purpose — regenerate and commit, and expect the
/// relay's own conformance test to fail until it is taught the same thing —
/// or it changed by accident, which is the case this exists for.
library;

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:broadcast/broadcast.dart';
import 'package:test/test.dart';

Uint8List _bytes(String hex) => hexDecode(hex);

void main() {
  late Map<String, dynamic> vectors;

  setUpAll(() {
    final file = File('test/conformance/wire_vectors.json');
    expect(
      file.existsSync(),
      isTrue,
      reason: 'run: dart run tool/emit_conformance_vectors.dart',
    );
    vectors = json.decode(file.readAsStringSync()) as Map<String, dynamic>;
  });

  test('the vectors are for a format version this build speaks', () {
    expect(vectors['version'], 1);
  });

  test('the fixed seeds still produce the recorded keys', () async {
    // If this fails, the key derivation changed and every recorded byte
    // below is meaningless — so it is checked first.
    final keys = vectors['keys'] as Map<String, dynamic>;
    final root = await CryptographyBroadcastSigner.fromSeed(
      _bytes(keys['rootSeedHex'] as String),
    );
    expect(hexEncode(root.publicKey), keys['rootPublicKeyHex']);
    expect(hexEncode(authorIdFor(root.publicKey)), keys['authorIdHex']);

    final publishing = await CryptographyBroadcastSigner.fromSeed(
      _bytes(keys['publishingSeedHex'] as String),
    );
    expect(hexEncode(publishing.publicKey), keys['publishingPublicKeyHex']);
  });

  test('the recorded certificate parses and verifies', () async {
    final keys = vectors['keys'] as Map<String, dynamic>;
    final record = vectors['certificate'] as Map<String, dynamic>;
    final encoded = _bytes(record['encodedHex'] as String);

    expect(encoded.length, record['byteLength']);
    expect(encoded.length, certificateBytes);

    final parsed = await PublishingKeyCertificate.verify(
      encoded: encoded,
      rootPublicKey: _bytes(keys['rootPublicKeyHex'] as String),
      verifier: const CryptographyBroadcastVerifier(),
      now: DateTime.fromMillisecondsSinceEpoch(
        (record['notBeforeSeconds'] as int) * 1000,
        isUtc: true,
      ),
    );
    expect(parsed, isNotNull);
    expect(hexEncode(parsed!.publishingKey), keys['publishingPublicKeyHex']);
    expect(
      parsed.notAfter.millisecondsSinceEpoch ~/ 1000,
      record['notAfterSeconds'],
    );
  });

  test(
    'every recorded descriptor parses to exactly what was recorded',
    () async {
      final keys = vectors['keys'] as Map<String, dynamic>;
      final rootKey = _bytes(keys['rootPublicKeyHex'] as String);
      final publishingKey = _bytes(keys['publishingPublicKeyHex'] as String);

      for (final entry in vectors['descriptors'] as List<dynamic>) {
        final record = entry as Map<String, dynamic>;
        final encoded = _bytes(record['encodedHex'] as String);
        final name = record['name'];

        expect(encoded.length, record['byteLength'], reason: '$name length');

        final verified = await BroadcastDescriptor.verify(
          encoded: encoded,
          rootPublicKey: rootKey,
          publishingKey: publishingKey,
          verifier: const CryptographyBroadcastVerifier(),
        );
        expect(verified, isNotNull, reason: '$name must verify');
        expect(verified!.seq, record['seq'], reason: '$name seq');
        expect(verified.flags, record['flags'], reason: '$name flags');
        expect(
          verified.publishedAt.millisecondsSinceEpoch ~/ 1000,
          record['publishedAtSeconds'],
          reason: '$name time',
        );
        expect(hexEncode(verified.id), record['idHex'], reason: '$name id');
        if (record.containsKey('prevHex')) {
          expect(hexEncode(verified.prev), record['prevHex']);
        }
        if (record.containsKey('retractsHex')) {
          expect(
            verified.isRetraction,
            isTrue,
            reason: '$name is a retraction',
          );
          expect(hexEncode(verified.retracts!), record['retractsHex']);
        } else {
          expect(verified.retracts, isNull, reason: '$name withdraws nothing');
        }
      }
    },
  );

  test(
    'the recorded relay directory verifies with its origins intact',
    () async {
      final keys = vectors['keys'] as Map<String, dynamic>;
      final record = vectors['relayDirectory'] as Map<String, dynamic>;
      final parsed = await RelayDirectory.verify(
        encoded: _bytes(record['encodedHex'] as String),
        rootPublicKey: _bytes(keys['rootPublicKeyHex'] as String),
        verifier: const CryptographyBroadcastVerifier(),
        now: DateTime.utc(2026, 2, 1),
      );
      expect(parsed, isNotNull);
      expect(parsed!.seq, record['seq']);
      expect(
        parsed.origins.map((u) => u.toString()).toList(),
        record['origins'],
      );
    },
  );

  test('the recorded bootstrap code parses from text and from bytes', () {
    final record = vectors['bootstrapCode'] as Map<String, dynamic>;
    final fromText = BootstrapCode.parse(record['text'] as String);
    final fromBytes = BootstrapCode.decodeBytes(
      _bytes(record['encodedHex'] as String),
    );
    expect(fromText, isNotNull);
    expect(fromText, fromBytes);
    expect(fromText!.host, record['host']);
    expect(
      hexEncode(fromText.rootPublicKey),
      (vectors['keys'] as Map<String, dynamic>)['rootPublicKeyHex'],
    );
  });

  test('the recorded addresses are the ones this build builds', () {
    final keys = vectors['keys'] as Map<String, dynamic>;
    final record = vectors['addresses'] as Map<String, dynamic>;
    final authorId = _bytes(keys['authorIdHex'] as String);

    expect(
      DescriptorAddress(authorId: authorId, seq: 41).path,
      record['descriptorPath'],
    );
    expect(
      ObjectAddress(contentHash(Uint8List.fromList([1, 2, 3]))).path,
      record['objectPath'],
    );
    expect(immutableCacheControl, record['immutableCacheControl']);
    // And they parse back, which is the property a relay depends on.
    expect(
      DescriptorAddress.tryParse(record['descriptorPath'] as String),
      isNotNull,
    );
    expect(ObjectAddress.tryParse(record['objectPath'] as String), isNotNull);
  });

  test('the recorded auth header is what this build sends', () async {
    final keys = vectors['keys'] as Map<String, dynamic>;
    final record = vectors['authHeader'] as Map<String, dynamic>;
    expect(broadcastAuthHeader, record['name']);

    final credentials = BroadcastCredentials(
      rootPublicKey: _bytes(keys['rootPublicKeyHex'] as String),
      certificate: _bytes(
        (vectors['certificate'] as Map<String, dynamic>)['encodedHex']
            as String,
      ),
    );
    expect(credentials.headerValue, record['value']);
  });

  test('regenerating the vectors would produce the same bytes', () async {
    // The generator is deterministic, so a run that changed the file would
    // mean the format moved. Checked here rather than trusted, because a
    // vector file that quietly drifts is worse than none.
    final keys = vectors['keys'] as Map<String, dynamic>;
    final publishing = await CryptographyBroadcastSigner.fromSeed(
      _bytes(keys['publishingSeedHex'] as String),
    );
    final rebuilt = await BroadcastDescriptor.sign(
      signer: publishing,
      authorId: _bytes(keys['authorIdHex'] as String),
      seq: 0,
      publishedAt: DateTime.utc(2026, 1, 1, 12),
      prev: zeroHash,
      layers: {
        LayerFlag.text: contentHash(Uint8List.fromList([1, 2, 3])),
      },
    );
    final recorded =
        (vectors['descriptors'] as List<dynamic>).first as Map<String, dynamic>;
    expect(hexEncode(rebuilt.encoded), recorded['encodedHex']);
  });
}

File: tools/cloudflare_relay_worker/test/conformance.test.mjs
/**
 * The relay's side of the shared wire vectors.
 *
 * The Dart client generates `packages/broadcast/test/conformance/
 * wire_vectors.json` and checks itself against it; this checks the relay
 * against the very same bytes. Two implementations of one format drift
 * silently otherwise — nothing fails until a deployment mixes an old
 * reader with a new writer, which is the one moment nobody can debug it.
 *
 * What is verified here is specifically what the relay depends on: that it
 * parses the client's paths, finds the author id at the offset it expects
 * inside a real descriptor, and accepts a real credential blob built by
 * the client's own encoder.
 */

import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { test } from 'node:test';

import {
  AUTH_BYTES,
  BroadcastArchive,
  authorizeDescriptorWrite,
  parseBroadcastPath,
  sha256Hex,
} from '../src/broadcast.js';

const VECTORS = JSON.parse(
  readFileSync(
    new URL(
      '../../../packages/broadcast/test/conformance/wire_vectors.json',
      import.meta.url,
    ),
    'utf8',
  ),
);

function fromHex(hex) {
  return new Uint8Array(hex.match(/../g).map((pair) => parseInt(pair, 16)));
}

function toBase64Url(bytes) {
  let binary = '';
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

class FakeStorage {
  constructor() {
    this.map = new Map();
  }

  async get(key) {
    return this.map.get(key);
  }

  async put(key, value) {
    this.map.set(key, value);
  }

  async delete(key) {
    return this.map.delete(key);
  }

  async list() {
    return new Map(this.map);
  }
}

test('the vectors are a format version this relay speaks', () => {
  assert.equal(VECTORS.version, 1);
});

test('the author id is the truncated hash of the root key', async () => {
  // The relay derives this itself to authorize a write, so if the two
  // sides ever disagree about the derivation, every publish fails.
  const rootKey = fromHex(VECTORS.keys.rootPublicKeyHex);
  const hash = await sha256Hex(rootKey);
  assert.equal(hash.slice(0, 32), VECTORS.keys.authorIdHex);
});

test('the client paths are the paths this relay routes', () => {
  const descriptor = parseBroadcastPath(VECTORS.addresses.descriptorPath);
  assert.equal(descriptor?.kind, 'descriptor');
  assert.equal(descriptor.authorId, VECTORS.keys.authorIdHex);
  assert.equal(descriptor.seq, 41);
  assert.equal(descriptor.shard, `a:${VECTORS.keys.authorIdHex}`);

  const object = parseBroadcastPath(VECTORS.addresses.objectPath);
  assert.equal(object?.kind, 'object');
  assert.equal(object.shard, `o:${object.hash.slice(0, 2)}`);
});

test('the credential blob the client sends is the size this relay expects',
  () => {
    const blob = fromHex(VECTORS.keys.rootPublicKeyHex + VECTORS.certificate.encodedHex);
    assert.equal(blob.length, AUTH_BYTES);
    assert.equal(toBase64Url(blob), VECTORS.authHeader.value);
    assert.equal(VECTORS.authHeader.name, 'x-broadcast-auth');
  });

test('a real descriptor and credentials authorize a real write', async () => {
  // End to end across both languages: bytes the Dart encoder produced,
  // verified by the relay's own Ed25519 path with no stub anywhere.
  for (const record of VECTORS.descriptors) {
    const authorized = await authorizeDescriptorWrite({
      authorId: VECTORS.keys.authorIdHex,
      descriptor: fromHex(record.encodedHex),
      authHeader: VECTORS.authHeader.value,
    });
    assert.equal(authorized, true, `${record.name} must authorize`);
  }
});

test('the recorded descriptor lengths match the format the relay caps', () => {
  for (const record of VECTORS.descriptors) {
    assert.equal(fromHex(record.encodedHex).length, record.byteLength);
    // The relay's descriptor ceiling has to leave room for the largest
    // real descriptor, or a legitimate publish is refused.
    assert.ok(record.byteLength <= 512, `${record.name} fits the cap`);
  }
});

test('a real post is stored and read back byte for byte', async () => {
  const archive = new BroadcastArchive(new FakeStorage());
  const record = VECTORS.descriptors[1];
  const bytes = fromHex(record.encodedHex);
  const route = parseBroadcastPath(VECTORS.addresses.descriptorPath);

  const write = await archive.handle(
    route,
    'PUT',
    bytes,
    VECTORS.authHeader.value,
  );
  assert.equal(write.status, 201);

  const read = await archive.handle(route, 'GET', null);
  assert.equal(read.status, 200);
  assert.deepEqual(new Uint8Array(await read.arrayBuffer()), bytes);
  assert.equal(
    read.headers.get('cache-control'),
    VECTORS.addresses.immutableCacheControl,
  );
});

test('a descriptor from these vectors cannot be written to another author',
  async () => {
    const archive = new BroadcastArchive(new FakeStorage());
    const otherAuthor = 'aabbccddeeff0011aabbccddeeff0011';
    const response = await archive.handle(
      { kind: 'descriptor', authorId: otherAuthor, seq: 41, shard: `a:${otherAuthor}` },
      'PUT',
      fromHex(VECTORS.descriptors[1].encodedHex),
      VECTORS.authHeader.value,
    );
    assert.equal(response.status, 403);
  });
